plugins/obs-transitions/data/luma_wipe_transition.effect

```diff
@@ -35,22 +37,31 @@ float4 PSLumaWipe(VertData v_in) : TARGET
 	float2 uv = v_in.uv;
 	float4 a_color = convert_pmalpha(a_tex.Sample(textureSampler, uv));
 	float4 b_color = convert_pmalpha(b_tex.Sample(textureSampler, uv));
-	float luma     = l_tex.Sample(textureSampler, uv).x;
+	float luma = l_tex.Sample(textureSampler, uv).x;
+	float soft = softness;
+
+	if (soft == 0.00)
+		soft += 0.01;
```

Comment:
Consider using clamp or min/max instead of if. Newer drivers _might_ optimize it into the same instructions, but older drivers will suffer from ifs that rely on uniforms.

Reply:
clamp would be better

Reply:
The problem is that if you use uniforms anywhere in the calculation, even anywhere earlier, and don't do some limiting on the uniform variable so the shader compiler has a known bound, you will get the same compilation route as just directly using the uniform. So these two have different compilation paths:
```
uniform float softness;
float soft = softness;
if (soft == 0.00) {
    soft += 0.001;
}
```
```
uniform float softness;
float soft = max(softness, 1.0e-6);
```
The first one will simply be an if, or perhaps converted to a more optimal check against 0, while the second one actually places a lower bound on soft (but no upper bound). Having any restrictive bounds can lead to better optimized final shader code than having no bounds. Though in this case it might not really make a difference on modern cards, but older cards, especially those struggling with ifs and jumps will see a difference.

Can softness be negative, or just a positive value of 0 to infinity?
```
uniform float softness;
float soft = max(softness, 1.0e-6);
```
Or is it -infinity to +infinity? In this case, abs and sign can be used.
```
uniform float softness;
float soft = max(abs(softness), 1.0e-6) * sign(softness);
```

Ideally this limiting would not be done on the GPU, as it can be done on the CPU at much lower cost.

Reply:
Thanks for the explanation. I was under the impression that converting variables away from uniform was enough to mitigate some of these issues.

I do like the max * sign, am going to test and probably implement.

Reply:
Consider simply doing this on the CPU, or at least caching the result in the vertex shader. Doing this calculation per fragment is going to relatively kill performance.

Reply:
Are you referring to the soft definition or the entire operation. On average the operation only affect a narrow range of pixels equal to about twice the softness range.

Reply:
I'm referring to the limiting of softness to not be 0, which is not a necessary per-fragment or per-vertex operation. The fragment shader is executed for every fragment, which means that your limiting happens 124416000 times per second at 1920x1080@60fps, compared to just 60 times if you did it on the CPU. That is rather huge difference that actually matters in situations where the GPU is used heavily by game.

Reply:
I agree, the placement of that calculation does not need to be per fragment. Making an adjustment

Reply:
Moved into the interface, not allowing 0 and only using positive numbers fir softness